src/parser.hpp

```diff
@@ -390,41 +390,31 @@ namespace wpp {
 			lex.advance();  // Skip `(`.
 
 
-			// Check for the first parameter.
-			if (lex.peek() == TOKEN_IDENTIFIER)
-				tree.get<Fn>(node).parameters.emplace_back(lex.advance().str());
-
-			else if (peek_is_reserved_name(lex.peek()))
-				throw wpp::Exception{lex.position(), "parameter name '", lex.advance().str(), "' conflicts with keyword."};
-
-
-			// While there is a comma, loop until we run out of parameters.
-			while (lex.peek() == TOKEN_COMMA) {
-				lex.advance(); // Skip `,`.
-
-				// Allow trailing comma.
-				if (lex.peek() == TOKEN_RPAREN)
-					break;
+			// While there is an identifier there is another parameter.
+			while (lex.peek() == TOKEN_IDENTIFIER) {
+				// Advance the lexer and get the identifier.
+				auto id = lex.advance().str();
 
+				// Add the argument
+				tree.get<Fn>(node).parameters.emplace_back(id);
 
-				// Check if there's an keyword conflict.
-				if (peek_is_reserved_name(lex.peek()))
-					throw wpp::Exception{lex.position(), "parameter name '", lex.advance().str(), "' conflicts with keyword."};
+				// If the next token is a comma, skip it.
+				if (lex.peek() == TOKEN_COMMA)
+					lex.advance(); // skip the comma
 
-
-				// Check if there is a parameter.
-				if (lex.peek() != TOKEN_IDENTIFIER)
-					throw wpp::Exception{lex.position(), "expecting parameter name to follow comma."};
-
-
-				// Emplace the parameter name into our `Fn` node's list of params.
-				tree.get<Fn>(node).parameters.emplace_back(lex.advance().str());
+				// Otherwise it must be an RPAREN?
+				else if (lex.peek() != TOKEN_RPAREN)
+					// If it's not, throw an exception.
+					throw wpp::Exception{lex.position(), "expecting comma to follow parameter name."};
 			}
 
+			// Check if there's an keyword conflict.
+			if (peek_is_reserved_name(lex.peek()))
```

Comment:
Not really, since we break out of the while loop if `lex.peek()` is not an identifier, and `peek_is_reserved_name` will always return false for identifiers. Basically it makes no sense to check every time because the moment we hit a keyword we automatically break out of the while loop, and then we check for the reason for the breakage. If it was a reserved keyword, we error appropriately, and same if it wasn't a closing parenthesis. In fact, that check for RPAREN could be an else if, but that would just make it confusing.

Reply:
Oh right I see now, could you maybe add a comment for that just to make it more clear?

Reply:
Sure

Reply:
What about something like this?
```cpp
			// We don't check this inside the while loop because for peek_is_reserved_name to return true,
			// it requires lex.peek() to be not an identifier, and the while loop ends when we run out of
			// identifiers. This section can be thought of as a check for the "reason" of breaking out of
			// the while loop (i.e. whether we reached a ')', or we found a reserved keyword).
```

Reply:
I think it's a bit waffley. Maybe something more like: 
```
We check if the next token is a reserved name and throw an error 
if it is. The reason we don't check this in the while loop body is 
because the loop condition checks for an identifier and so breaks
out if the next token is an intrinsic. 
```